File: jianzhioffer/firstround/jianzhi_52.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        node1 = headA
        node2 = headB

        while node1 != node2:
            node1 = node1.next if node1 else headB
            node2 = node2.next if node2 else headA
        return node1


if __name__ == '__main__':
    l11 = ListNode(1)
    l12 = ListNode(2)
    l13 = ListNode(3)
    l14 = ListNode(4)
    #l5 = ListNode(5)
    l11.next = l12
    l12.next = l13
    l13.next = l14
    #l4.next = l5

    l21 = ListNode(-1)

    obj = Solution()
    node = obj.getIntersectionNode(l11, l21)
    print(node.val)

File: jianzhioffer/secondround/jianzhi_52.cpp
#include <cstddef>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x): val(x), next(NULL) {};

};

class Solution
{
public:
    Solution();
    ~Solution();
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB);
};

Solution::Solution() {}

Solution::~Solution() {}

//时间复杂度O(M+N)，空间复杂度O(1)
ListNode* Solution::getIntersectionNode(ListNode* headA, ListNode* headB)
{
    if(headA==NULL || headB==NULL) return NULL;
    ListNode* curA = headA;
    ListNode* curB = headB;
    while (curA!=curB)
    {
        if(curA!=NULL) curA = curA->next;
        else curA = headB;

        if(curB!=NULL) curB = curB->next;
        else curB = headA;
    }
    return curA;
}


int main(int argc, char const *argv[])
{
    ListNode l11 = ListNode(4);
    ListNode l12 = ListNode(1);
    ListNode l21 = ListNode(5);
    ListNode l3 = ListNode(8);
    ListNode l4 = ListNode(4);
    l11.next = &l12;
    l12.next = &l3;
    // l21.next = &l3;
    l3.next = &l4;

    Solution obj;
    ListNode* result = obj.getIntersectionNode(&l11, &l21);
    return 0;
}
